### PR TITLE
Master variant view yti

### PR DESCRIPTION
**Pull request by tivisse**
TASK : https://www.odoo.com/web#id=10789&view_type=form&model=project.task&action=333&active_id=131
PAD : https://pad.odoo.com/p/r.DJtezlEkOevRWOlm

### CODE REVIEW

**Comment by rim-odoo** [comment]
@tivisse Conflicts spotted, a good rebase could help ;-)

File: addons/base_setup/res_config_view.xml [hunk 1]
```diff
@@ -19,7 +19,7 @@
                             <button type="object" name="open_company" string="Configure your company data" icon="gtk-execute" class="oe_inline oe_link"/>
                         </p>
                     </div>
-                    <group>
+                    <group name="options">
```
Comment (tde-banana-odoo) [inline addons/base_setup/res_config_view.xml]: Why adding a name ? It does not seem to be used in this diff.

File: addons/product/product.py [hunk 2]
```diff
@@ -470,6 +470,74 @@ def _get_product_variant_count(self, cr, uid, ids, name, arg, context=None):
             res[product.id] = len(product.product_variant_ids)
         return res
 
+    def _get_product_template_standard_price(self, cr, uid, ids, name, arg, context=None):
```
Comment (tde-banana-odoo) [inline addons/product/product.py]: Seems like there is place to have a generic method here, taking the field name as argument.

And also maybe we can batch
res = dict.fromkey(ids, 0.0)
unique_templates = [template for template in self.browse() if template.variant_count == 1]
res.update(dict((template.product_variant_ids[0].id, template.product_variant_ids[0].standard_price) for template in unique_templates))

File: addons/product/product.py [hunk 3]
```diff
@@ -470,6 +470,74 @@ def _get_product_variant_count(self, cr, uid, ids, name, arg, context=None):
...
+        ''' Compute the standard price from the product_variant if there is only one variant, otherwise returns 0.0 '''
+        res = {}
+        for template in self.browse(cr, uid, ids, context=context):
+            res[template.id] = self.pool['product.product'].browse(cr, uid, template.product_variant_ids.id, context=context).standard_price if template.product_variant_count == 1 else 0.0
+        return res
+
+    def _set_product_template_standard_price(self, cr, uid, product_tmpl_id , name, value, args, context=None):
```
Comment (tde-banana-odoo) [inline addons/product/product.py]: Probably factorize bulk of the method.

File: addons/product/product.py [hunk 4]
```diff
@@ -488,13 +556,11 @@ def _get_product_variant_count(self, cr, uid, ids, name, arg, context=None):
...
-                                          help="Cost price of the product template used for standard stock valuation in accounting and used as a base price on purchase orders. "
-                                               "Expressed in the default unit of measure of the product.",
-                                          string="Cost Price"),
-        'volume': fields.float('Volume', help="Volume is the amount of space that an item you are measuring takes up."),
-        'weight': fields.float('Gross Weight', digits_compute=dp.get_precision('Stock Weight'), help="The total weight, including contents, packaging, etc."),
-        'weight_net': fields.float('Net Weight', digits_compute=dp.get_precision('Stock Weight'), help="The weight of the contents, not including any packaging, etc."),
+        'standard_price': fields.function(_get_product_template_standard_price, fnct_inv=_set_product_template_standard_price ,type='float', string='Cost Price',digits_compute=dp.get_precision('Product Price'), help="Cost price of the product template used for standard stock valuation in accounting and used as a base price on purchase orders. "
```
Comment (tde-banana-odoo) [inline addons/product/product.py]: Just to ease migration: _compute_<field_name> and _set_<field_name>

File: addons/product/product_view.xml [hunk 5]
```diff
@@ -211,17 +211,66 @@
               </p>
             </field>
         </record>
+
+        <record id="product_variant_easy_edit_view" model="ir.ui.view">
+            <field name="name">product.product.easy.edit.form</field>
```
Comment (tde-banana-odoo) [inline addons/product/product_view.xml]: product.product.view.form.light (or easy, or anything cool and clean)

File: addons/product/product.py (hunk 2 above)
Comment (tivisse) [inline addons/product/product.py]: Gweat Idea !

File: addons/product/product_view.xml [hunk 6]
```diff
@@ -211,17 +211,66 @@
...
             <field name="res_model">product.product</field>
-            <field name="view_mode">tree,form,kanban</field>
-            <field name="view_type">form</field>
             <field name="context">{'search_default_product_tmpl_id': [active_id], 'default_product_tmpl_id': active_id}</field>
-            <field name="search_view_id" ref="product_search_form_view"/>
-            <field name="view_id" eval="False"/> <!-- Force empty -->
-            <field name="domain" eval="False"/> <!-- Force empty -->
-            <field name="help" type="html">
+            <field name="search_view_id" ref="product_search_form_view"/>             
+            <field name="view_ids"
```
Comment (tde-banana-odoo) [inline addons/product/product_view.xml]: Seems nicer than defining unnecessary xml_ids. Al and I approve.

Just add spaces :postal_horn:  !

File: addons/product/product_view.xml [hunk 7]
```diff
@@ -236,8 +285,7 @@
             </field>
         </record>
 
-
-        <!-- view specific to product.template -->
+       <!-- view specific to product.template -->
```
Comment (tde-banana-odoo) [inline addons/product/product_view.xml]: Unnecessary (and wrong) diff.

File: addons/product/product_view.xml [hunk 8]
```diff
@@ -252,21 +300,25 @@
                     <field name="barcode" attrs="{'invisible': [('product_variant_count', '>', 1)]}"/>
                     <field name="default_code" attrs="{'invisible': [('product_variant_count', '>', 1)]}"/>
                 </field>
+
                 <div name="button_box" position="inside">
-                    <button class="oe_stat_button" string="Variant Prices"  name="%(variants_template_action)d" type="action" icon="fa-usd"/>
-                    <button class="oe_stat_button" name="%(product.product_variant_action)d" type="action" icon="fa-sitemap">
+                    <button class="oe_stat_button" string="Variant Prices"  name="%(variants_template_action)d" type="action" icon="fa-usd" groups="product.group_product_variant"/>
+                    <button class="oe_stat_button" name="%(product.product_variant_action)d" type="action" icon="fa-sitemap" groups="product.group_product_variant"/>
+                    <button class="oe_stat_button" string="Variant Prices"  name="%(variants_template_action)d" type="action" icon="fa-strikethrough" groups="product.group_product_variant"/>
```
Comment (tde-banana-odoo) [inline addons/product/product_view.xml]: Isn't it the same than 2 buttons above, except the icon ?

File: addons/product/product_view.xml [hunk 9]
```diff
@@ -252,21 +300,25 @@
...
+                    <button class="oe_stat_button" name="%(product.product_variant_action)d" type="action" icon="fa-sitemap" groups="product.group_product_variant">
```
Comment (tde-banana-odoo) [inline addons/product/product_view.xml]: Isn't it the same than 2 buttons above, except the icon ?

File: addons/stock/test/packing.yml [hunk 10]
```diff
@@ -121,9 +121,9 @@
     reco_id = self.search(cr ,uid , [('product_id','=',ref('product1'))], context=context)
     for rec in self.browse(cr, uid, reco_id, context=context):
       if rec.package_id.name == 'Pallet 1' and rec.location_id.id == ref('stock_location_customers'):
-        assert rec.qty == 120, "Should have 120 pieces on pallet 1"
+        assert rec.qty == 120, "Should have 120 pieces on pallet 1, got " +  str(rec.qty)
```
Comment (tde-banana-odoo) [inline addons/stock/test/packing.yml]: Do a specific commit for this file (to avoid "by the way" changes).

File: addons/product/product.py [hunk 11]
```diff
@@ -488,13 +505,12 @@ def _get_product_variant_count(self, cr, uid, ids, name, arg, context=None):
         'price': fields.function(_product_template_price, fnct_inv=_set_product_template_price, type='float', string='Price', digits_compute=dp.get_precision('Product Price')),
         'list_price': fields.float('Sale Price', digits_compute=dp.get_precision('Product Price'), help="Base price to compute the customer price. Sometimes called the catalog price."),
         'lst_price' : fields.related('list_price', type="float", string='Public Price', digits_compute=dp.get_precision('Product Price')),
-        'standard_price': fields.property(type = 'float', digits_compute=dp.get_precision('Product Price'), 
+        'standard_price': fields.function(_compute_product_template_field, fnct_inv=_set_product_template_field, type='float', string='Cost Price', digits_compute=dp.get_precision('Product Price'),
```
Comment (tde-banana-odoo) [inline addons/product/product.py]: While you are at it, could'nt you make a "multi" field ?

Is it normal that it is not stored ?

File: addons/product/res_config.py [hunk 12]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (tde-banana-odoo) [inline addons/product/res_config.py]: NotImportant: not necessary diff

File: addons/product/__init__.py [hunk 13]
```diff
@@ -6,3 +6,4 @@
 import report
 import partner
 import wizard
+import res_config
```
Comment (tde-banana-odoo) [inline addons/product/__init__.py]: Useless import.

File: addons/product/product.py [hunk 14]
```diff
@@ -470,6 +470,27 @@ def _get_product_variant_count(self, cr, uid, ids, name, arg, context=None):
             res[product.id] = len(product.product_variant_ids)
         return res
 
+    def _compute_product_template_field(self, cr, uid, ids, names, arg, context=None):
+        ''' Compute the field from the product_variant if there is only one variant, otherwise returns 0.0 '''
+        import pdb; pdb.set_trace()
```
Comment (tde-banana-odoo) [inline addons/product/product.py]: Is the branch black ?

File: addons/product/product.py [hunk 15]
```diff
@@ -470,6 +470,27 @@ def _get_product_variant_count(self, cr, uid, ids, name, arg, context=None):
...
+        templates = [template for template in self.browse(cr, uid, ids, context=context)]
+        unique_templates = [template.id for template in templates if template.product_variant_count == 1]
+        for template in templates:
+            for name in names:
+                res[template.id][name] = getattr(template.product_variant_ids[0], name) if template.id in unique_templates else 0.0
+        return res     
+
+    def _set_product_template_field(self, cr, uid, product_tmpl_id, name, value, args, context=None):
+        ''' Set the standard price modification on the variant if there is only one variant '''
+        if context is None:
```
Comment (tde-banana-odoo) [inline addons/product/product.py]: As you are not actively using the context, it is not necessary to check its existence.

**Comment by tivisse** [comment]
Landed in master https://github.com/odoo/odoo/commit/3befe66f432d02178512ba52c6d9fdfc314e9ff9

**Pull request merged**